File: OpenMP/answers/ex4_dotprod.c
#include <stdio.h>

#define NX 102400

int main(void)
{
    double vecA[NX], vecB[NX];
    double sum, psum;
    int i;

    /* Initialization of the vectors */
    for (i = 0; i < NX; i++) {
        vecA[i] = 1.0 / ((double) (NX - i));
        vecB[i] = vecA[i] * vecA[i];
    }

    sum = 0.0;
    /* Version with data race */
#pragma omp parallel for default(shared) private(i)
    for (i = 0; i < NX; i++) {
        sum += vecA[i] * vecB[i];
    }
    printf("Sum with data race:                              %18.16f\n",
           sum);

    sum = 0.0;
    /* Dot product using critical section = SERIAL CODE! */
#pragma omp parallel for default(shared) private(i)
    for (i = 0; i < NX; i++) {
#pragma omp critical(dummy)
        sum += vecA[i] * vecB[i];
    }
    printf("Sum using critical section:                      %18.16f\n",
           sum);

    sum = 0.0;
    /* Dot product using private partial sums and critical section */
#pragma omp parallel default(shared) private(i, psum)
    {
        psum = 0.0;
#pragma omp for
        for (i = 0; i < NX; i++) {
            psum += vecA[i] * vecB[i];
        }
#pragma omp critical(par_sum)
        sum += psum;
    }
    printf("Sum using private variable and critical section: %18.16f\n",
           sum);

    sum = 0.0;
    /* Dot product using reduction */
#pragma omp parallel for default(shared) private(i) reduction(+:sum)
    for (i = 0; i < NX; i++) {
        sum += vecA[i] * vecB[i];
    }
    printf("Reduction sum:                                   %18.16f\n",
           sum);

    return 0;
}

File: OpenMP/answers/ex4_dotprod.f90
program dotprod
  implicit none
  integer, parameter :: rk = kind(1d0)
  integer, parameter :: ik = selected_int_kind(9)
  integer, parameter :: nx = 102400

  real(kind=rk), dimension(nx) :: vecA, vecB
  real(kind=rk)    :: sum, psum
  integer(kind=ik) :: i

  ! Initialization of vectors
  do i = 1, nx
     vecA(i) = 1.0_rk/(real(nx - i + 1, kind=rk))
     vecB(i) = vecA(i)**2
  end do

  sum = 0.0_rk
  ! Sum with data race
  !$omp parallel do default(shared) private(i)
  do i = 1, nx
     sum = sum + vecA(i) * vecB(i)
  end do
  !$omp end parallel do
  write(*,*) 'Sum using critical section:                      ', sum

  sum = 0.0_rk
  ! Dot product using critical section = SERIAL CODE
  !$omp parallel do default(shared) private(i)
  do i = 1, nx
     !$omp critical
     sum = sum + vecA(i) * vecB(i)
     !$omp end critical
  end do
  !$omp end parallel do
  write(*,*) 'Sum using critical section:                      ', sum

  sum = 0.0_rk
  ! Dot product using reduction
  !$omp parallel do default(shared) private(i) reduction(+:sum)
  do i = 1, nx
     !$omp critical
     sum = sum + vecA(i) * vecB(i)
     !$omp end critical
  end do
  !$omp end parallel do
  write(*,*) 'Sum using reduction:                             ', sum

  sum = 0.0_rk
  ! Dot product using private variable and critical secion
  !$omp parallel default(shared) private(i, psum)
  psum = 0.0_rk
  !$omp do
  do i = 1, nx
     psum = psum + vecA(i) * vecB(i)
  end do
  !$omp end do
  !$omp critical
  sum = sum + psum
  !$omp end critical
  !$omp end parallel
  write(*,*) 'Sum using private variable and critical section: ', sum

end program dotprod
